File: 01-javaScript/phase-1.js
// 1. FizzBuzz

function fizzBuzz(n) {
    for(let i = 1; i <= n; i++){

        if(i % 3 === 0 && i % 5 === 0){
            console.log("FizzFuzz")
        }
        else if(i % 3 === 0){
            console.log("Fizz")
        }
        else if(i % 5 === 0){
            console.log("Fuzz")
        }
        else {
            print(i)
        }
    }
}

// fizzBuzz(100)





///////////////////////////////////////////////////////////////////////////////
// Challenge 2: String Reversal

function stringReversal(text=""){
    let reversedString = ""

    // for(let i = 1; i <= text.length ; i++){
    //     reversedString += text[text.length - i]
    // }
    for(let i = text.length - 1; i >= 0 ; i--){
        reversedString += text[i]
    }

    return reversedString
}

// console.log(stringReversal("Sameer"))


///////////////////////////////////////////////////////////////////////////////
// Challenge 3 Palindrome Checker

function palindromeChecker(word=""){
    let cleanWord = word.toLowerCase().split(" ").join('')
    let lastIndex = cleanWord.length - 1
    let isPalindrome = false

    for(let i = 0 ; i < Math.round(cleanWord.length / 2); i++){
        if(cleanWord[i] !== cleanWord[lastIndex]){
            return "this is not a palindrome word"
        }
        isPalindrome = true
        lastIndex--
    }

    if(isPalindrome === true){
        return "This a palindrome word"
    }
}



// AI's Answer

// function isPalindrome(text) {
//     // Step 1: Regex ka use karke a-z aur 0-9 ke alawa sab kuch ('') empty string se replace kar do
//     // /[^a-z0-9]/gi ka matlab: g = global (puri string mein), i = case insensitive
//     const cleaned = text.replace(/[^a-z0-9]/gi, '').toLowerCase();
    
//     // Step 2: Two Pointers
//     let left = 0;
//     let right = cleaned.length - 1;
    
//     while (left < right) {
//         if (cleaned[left] !== cleaned[right]) {
//             return false;
//         }
//         left++;
//         right--;
//     }
    
//     return true;
// }

// console.log(palindromeChecker("pat t ap"))
// console.log(palindromeChecker("No lemon, no melon"))






///////////////////////////////////////////////////////////////////////////////
// Challenge 4 (Factorial - Loop vs Recursion)


// Loop
function factorial(n){
    sum = 1

    for(let i = 1; i < n; i++){
        // console.log(`${sum} * ${i}`)
        sum += sum * i
    }
    return sum
}

function factorialRecursive(n) {
    if (n < 0) return "Invalid input";
    
    // BASE CASE
    if (n === 0 || n === 1) {
        return 1;
    }
    
    // RECURSIVE CALL
    return n * factorialRecursive(n - 1);
}

// console.log(factorial(1))
// console.log(factorial(2))
// console.log(factorial(3))
// console.log(factorial(4))
// console.log(factorial(5))
// console.log(factorial(10))

File: 02-Python/phase-1.py
# FizzFuzz

# def fizz_fuzz(n=100):
#     for i in range(1, n + 1):
#         print(i)
#         if i % 3 == 0 and i % 5 == 0:
#             print("FizzFuzz")
#         elif i % 3 == 0:
#             print("Fizz")
#         elif i % 5 == 0:
#             print("Fuzz")
#         else :
#             print(i)


# fizz_fuzz()



#####################################################################################
# Challenge 2: String Reversal

# My Answer 

# def string_reverse(text=""):
#     reversedString = ""

#     for i in range(1, len(text) + 1):
#         print(i)
#         reversedString += text[len(text) - i]

#     return reversedString

# print(string_reverse("sameer"))


# AI's Answer

# def reverse_string(text):
#     reversed_str = ""
    
#     # Text ke har character par loop chalayenge
#     for char in text:
#         # Naye character ko purani string ke AAGE jod rahe hain
#         reversed_str = char + reversed_str 
        
#     return reversed_str

# print(reverse_string("hello"))




#####################################################################################
# Challenge 3 Palindrome Checker

def palindrome_checker(word=""):
    clean_word = word.lower().replace(' ', '')
    last_index = len(clean_word) - 1

    for char in clean_word:
        if char is not clean_word[last_index]:
            return "This is not a palindrome word"

        last_index -= 1
    else :
        return "This is a Palindrome word"



# AI's Answer

def is_palindrome(text):
    # Step 1: String ko clean karna (Sirf alphabets/numbers rakhna aur lowercase karna)
    # isalnum() check karta hai ki character alphabet ya number hai (spaces/symbols nahi)
    cleaned = ''.join(char.lower() for char in text if char.isalnum())
    
    # Step 2: Two Pointers initialize karna
    left = 0
    right = len(cleaned) - 1
    
    while left < right:
        # Agar characters match nahi karte, toh palindrome nahi hai
        if cleaned[left] != cleaned[right]:
            return False
            
        # Pointers ko move karna
        left += 1
        right -= 1
        
    return True


# print(palindrome_checker("pattap"))
# print(palindrome_checker("pat t ap"))
# print(palindrome_checker("123421"))
# print(palindrome_checker("123ab321"))
# print(palindrome_checker("No lemon, no melon"))






#####################################################################################
# Challenge 4 (Factorial - Loop vs Recursion)

def factorial(n):
    sum = 1

    for i in range(1,n):
        sum += i * sum

    return sum


def factorial_recursive(n):
    if n < 0:
        return "Invalid Input"
        
    # BASE CASE (Rukne ka point)
    if n == 0 or n == 1:
        return 1
        
    # RECURSIVE CALL
    return n * factorial_recursive(n - 1)



print(factorial(5))